### PR TITLE
OWLS-68700: Add support for pod labels and annotations

### PR DESCRIPTION
**Pull request by swapnanitin**

### CODE REVIEW

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java [hunk 1]
```diff
@@ -77,6 +79,26 @@
   @Description("Additional volume mounts")
   private List<V1VolumeMount> additionalVolumeMounts = new ArrayList<>();
 
+  /**
+   * The additional volumes.
```
Comment (vkraemer) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java]: are labels volume mounts?

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java [hunk 2]
```diff
@@ -77,6 +79,26 @@
...
+   *
+   * @since 2.0
+   */
+  @SerializedName("podLabels")
+  @Expose
+  @Description("Labels applied to pods")
+  private Map<String, String> podLabels = new HashMap<String, String>();
+
+  /**
+   * The additional volume mounts.
```
Comment (vkraemer) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java]: are annotation volume mounts?

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerSpec.java [hunk 3]
```diff
@@ -154,15 +154,19 @@ public ProbeTuning getReadinessProbe() {
...
-  @Nonnull
-  public Map<String, String> getPodLabels() {
-    return Collections.emptyMap();
-  }
+  /**
+   * Returns the labels applied to the pod.
+   *
+   * @return a map of labels
+   */
+  public abstract Map<String, String> getPodLabels();
```
Comment (vkraemer) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerSpec.java]: why did @Nonnull disappear here?

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerSpec.java [hunk 4]
```diff
@@ -154,15 +154,19 @@ public ProbeTuning getReadinessProbe() {
...
-  @Nonnull
-  public Map<String, String> getPodAnnotations() {
-    return Collections.emptyMap();
-  }
+  /**
+   * Returns the annotations applied to the pod.
+   *
+   * @return a map of annotations
+   */
+  public abstract Map<String, String> getPodAnnotations();
```
Comment (vkraemer) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerSpec.java]: same question.

File: operator/src/test/java/oracle/kubernetes/operator/helpers/AdminPodHelperTest.java [hunk 5]
```diff
@@ -328,6 +330,75 @@ public void whenPodHasDuplicateVolumeMounts_createAdminPodWithCombination() {
             hasVolumeMount("volume1", "/domain-path1"), hasVolumeMount("volume2", "/server-path")));
   }
 
+  @Test
```
Comment (vkraemer) [inline operator/src/test/java/oracle/kubernetes/operator/helpers/AdminPodHelperTest.java]: maybe add a test for the 'I did not add labels or annotations' cases, too.

File: operator/src/main/java/oracle/kubernetes/operator/helpers/PodStepContext.java [hunk 6]
```diff
@@ -569,6 +573,8 @@ protected V1ObjectMeta createMetadata() {
             .putLabelsItem(LabelConstants.DOMAINNAME_LABEL, getDomainName())
             .putLabelsItem(LabelConstants.SERVERNAME_LABEL, getServerName())
             .putLabelsItem(LabelConstants.CREATEDBYOPERATOR_LABEL, "true");
+    getPodLabels().forEach((k, v) -> metadata.putLabelsItem(k, v));
```
Comment (rjeberhard) [inline operator/src/main/java/oracle/kubernetes/operator/helpers/PodStepContext.java]: Looking at this, I'm realizing that we need validation that the customer doesn't try to create a label matching one of the labels that we own.  The easiest rule would be that the customer cannot create a label starting with "weblogic."  Thoughts?

File: operator/src/test/java/oracle/kubernetes/operator/helpers/AdminPodHelperTest.java (hunk 5 above)
Comment (swapnanitin) [reply operator/src/test/java/oracle/kubernetes/operator/helpers/AdminPodHelperTest.java]: We set internal labels and annotations so its never empty.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java [hunk 7]
```diff
@@ -187,6 +219,22 @@ private void addAdditionalVolumeMount(V1VolumeMount var) {
...
+  public Map<String, String> getPodLabels() {
+    return podLabels;
+  }
+
+  public Map<String, String> getPodAnnotations() {
+    return podAnnotations;
+  }
+
   @Override
   public String toString() {
```
Comment (AdrianPadilla) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java]: Maybe a good idea to add the new fields here

**Comment by AdrianPadilla** [comment]
New fields in ServerPod should be added to hashCode and equals methods

**Comment by rjeberhard** [comment]
Please add the new fields to toString, equals, and hashCode.  Then this is good to go.

**Pull request merged**